src/Test/StateMachine/Sequential.hs

```diff
@@ -193,69 +196,126 @@ getUsedVars :: Rank2.Foldable f => f Symbolic -> [Var]
+-- something like
+--
+-- > [    -- outermost list recording all the shrink possibilities
+-- >     [A', B1', C', D', E' , F', G', H']   -- B shrunk to B1
+-- >   , [A', B1', C', D', E' , F', G', H']   -- B shrunk to B2
+-- >   , [A', B' , C', D', E1', F', G', H']   -- E shrunk to E1
+-- >   , [A', B' , C', D', E2', F', G', H']   -- E shrunk to E2
+-- >   , [A', B' , C', D', E3', F', G', H']   -- E shrunk to E3
+-- > ]
+--
+-- where one of the commands has been shrunk and all commands have been
+-- validated and renumbered (references updated). So, in this example, the
+-- result will contain at most 5 lists; it may contain fewer, since some of
+-- these lists may not be valid.
+--
+-- If we _did_ already shrink something, then no commands will be shrunk, and
+-- the resulting list will either be empty (if the list of commands was invalid)
+-- or contain a /single/ element with the validated and renumbered commands.
+shrinkAndValidate :: forall model cmd m resp. (Rank2.Traversable cmd, Rank2.Foldable resp)
+                  => StateMachine model cmd m resp
+                  -> ShouldShrink
+                  -> ValidateEnv model
+                  -> Commands cmd resp
+                  -> [(ValidateEnv model, Commands cmd resp)]
+shrinkAndValidate StateMachine { precondition, transition, mock, shrinker } =
+    \env shouldShrink cmds -> map (second Commands) $ go env shouldShrink (unCommands cmds)
+  where
+    go :: ShouldShrink -> ValidateEnv model -> [Command cmd resp] -> [(ValidateEnv model, [Command cmd resp])]
+    go MustShrink   _   [] = []          -- we failed to shrink anything
+    go DontShrink   env [] = [(env, [])] -- successful termination
+    go shouldShrink (ValidateEnv model scope counter) (Command cmd' _resp vars' : cmds) =
       case Rank2.traverse (remapVars scope) cmd' of
-        Just cmd | boolean (precondition model cmd) -> do
-          let (resp, counter') = runGenSym (mock model cmd) counter
-              vars             = getUsedVars resp
-
-          put ( transition model cmd resp
-              , M.fromList (zip vars' vars) `M.union` scope
-              , counter')
-          mih <- go cmds
-          case mih of
-            Nothing -> return Nothing
-            Just ih -> return (Just (Command cmd vars : ih))
-        _otherwise ->
-          return Nothing
+        Just remapped ->
+          -- shrink at most one command
+          let candidates :: [(ShouldShrink, cmd Symbolic)]
+              candidates =
+                case shouldShrink of
+                  DontShrink -> [(DontShrink, remapped)]
+                  MustShrink -> map (DontShrink,) (shrinker model remapped)
+                             ++ [(MustShrink, remapped)]
+          in flip concatMap candidates $ \(shouldShrink', cmd) ->
+               if boolean (precondition model cmd)
+                 then let (resp, counter') = runGenSym (mock model cmd) counter
```

Comment:
Can we get rid the use of `mock` here as well?

Reply:
No, we cannot: since we are executing the command in a different model, it may return a different result, so we really need the `mock` function to get us a new response.